### PR TITLE
Add Simulation Setup

### PR DESCRIPTION
**Pull request by VishGit1234**
Add the following to the monorepo:
- CARLA and CARLA ROS Bridge
- CARLA Viz Tool
- Jupyter Notebook Server with CARLA dependencies
- Python sample node for CARLA

### CODE REVIEW

File: docker/simulation/carla_ros_bridge/carla_ros_bridge.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+RUN apt-get update && \
+      apt-fast install -qq -y --no-install-recommends lsb-release \
+      libglu1-mesa-dev xorg-dev \
+      software-properties-common \
+      build-essential \
+      wget \
+      curl \
+      python3-rosdep \
+      python3-rospkg \
+      git \
```
Comment (Edwardius) [inline docker/simulation/carla_ros_bridge/carla_ros_bridge.Dockerfile:53]: Specifically install this in the source stage (line 8), we do not want git installed in our final product image due to security reasons

File: docker/simulation/carla_sample_node/carla_sample_node.Dockerfile [hunk 2]
```diff
@@ -0,0 +1,63 @@
+ARG BASE_IMAGE=ghcr.io/watonomous/wato_monorepo/base:humble-ubuntu22.04
+
+################################ Source ################################
+FROM ${BASE_IMAGE} as source
+
+WORKDIR ${AMENT_WS}/src
+
+# Copy in source code 
+COPY src/simulation/carla_sample_node carla_sample_node
+COPY src/wato_msgs/sample_msgs sample_msgs
```
Comment (Edwardius) [inline docker/simulation/carla_sample_node/carla_sample_node.Dockerfile:10]: you might not need this anymore

File: docker/simulation/carla_sample_node/carla_sample_node.Dockerfile [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+COPY src/wato_msgs/simulation/common_msgs common_msgs
+COPY src/wato_msgs/simulation/embedded_msgs embedded_msgs
+COPY src/wato_msgs/simulation/path_planning_msgs path_planning_msgs
+
+# Carla specific messages
+RUN git clone --depth 1 https://github.com/ros-drivers/ackermann_msgs.git --branch ros2 && \
+    git clone --depth 1 https://github.com/ros-perception/image_common.git --branch $ROS_DISTRO && \
+    git clone --depth 1 https://github.com/carla-simulator/ros-carla-msgs.git --branch master
```
Comment (Edwardius) [inline docker/simulation/carla_sample_node/carla_sample_node.Dockerfile:18]: git is working here?

File: docker/simulation/carla_ros_bridge/carla_ros_bridge.Dockerfile [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+ARG CARLA_VERSION
+
+# Install dependencies
+RUN apt-get update && \
+      apt-fast install -qq -y --no-install-recommends lsb-release \
+      libglu1-mesa-dev xorg-dev \
+      software-properties-common \
+      build-essential \
+      wget \
+      curl \
```
Comment (Edwardius) [inline docker/simulation/carla_ros_bridge/carla_ros_bridge.Dockerfile:50]: if you can, either move these to the source stage (line 8) or remove. Security risk.

File: docker/simulation/carla_sim/carla_sim.Dockerfile [hunk 5]
```diff
@@ -1,55 +1,7 @@
-ARG BASE_IMAGE=ghcr.io/watonomous/wato_monorepo/base:humble-ubuntu22.04
+ARG CARLA_VERSION=0.9.13
+FROM carlasim/carla:${CARLA_VERSION} 
 
-################################ Source ################################
-FROM ${BASE_IMAGE} as source
+ARG CARLA_VERSION=0.9.13
```
Comment (Edwardius) [inline docker/simulation/carla_sim/carla_sim.Dockerfile:4]: dup?

File: docker/simulation/carla_sim/carla_sim.Dockerfile
Comment (Edwardius) [inline docker/simulation/carla_sim/carla_sim.Dockerfile:1]: this file might not be needed, I feel like you can do all of this in just the docker-compose file ( `image: carlasim/carla:${CARLA_VERSION:-?}` , the `?` will cause an error if the CARLA_VERSION is not set)

Comment (Edwardius) [reply docker/simulation/carla_sim/carla_sim.Dockerfile:1]: ofc, make sure its configurable with watod

File: docker/simulation/carla_viz/carla_viz.Dockerfile [hunk 6]
```diff
@@ -0,0 +1,75 @@
+FROM ubuntu:18.04 as repo
+RUN apt-get update -y && apt-get -y install git
+RUN git clone https://github.com/mjxu96/carlaviz.git && \
+    cd carlaviz && git checkout a8d6e66f4af91a7af5fe8169a03cc089c33ff62e
+WORKDIR /
+RUN git clone https://github.com/mjxu96/xviz.git && \
+    cd xviz && git checkout 6a45a56a7e62489d2200182a3379bb76fbf71141
```
Comment (Edwardius) [inline docker/simulation/carla_viz/carla_viz.Dockerfile:7]: Ideally, we wouldn't want git inside the container. You can separate git from the image by doing a multistage build
```suggestion
FROM ubuntu:18.04 as source
RUN apt-get update -y && apt-get -y install git
RUN git clone https://github.com/mjxu96/carlaviz.git && \
    cd carlaviz && git checkout a8d6e66f4af91a7af5fe8169a03cc089c33ff62e
WORKDIR /
RUN git clone https://github.com/mjxu96/xviz.git && \
    cd xviz && git checkout 6a45a56a7e62489d2200182a3379bb76fbf71141
   
FROM ubuntu:18.04 as repo
COPY --from=source carlaviz carlaviz
COPY --from=source xviz xviz
```

Comment (Edwardius) [reply docker/simulation/carla_viz/carla_viz.Dockerfile:7]: oh wait, it this separated out already?

Comment (Edwardius) [reply docker/simulation/carla_viz/carla_viz.Dockerfile:7]: oh yeah it is

Comment (Edwardius) [reply docker/simulation/carla_viz/carla_viz.Dockerfile:7]: idk, ill keep this in for an FYI

File: modules/dev_overrides/docker-compose.simulation.yaml [hunk 7]
```diff
@@ -6,10 +6,49 @@ x-fixuid: &fixuid
...
+  carla_sample_node:
+    <<: *fixuid
+    extends:
+      file: ../docker-compose.simulation.yaml
+      service: carla_sample_node
+    command: tail -F anything
+    volumes:
+      - ${MONO_DIR}/src/simulation/carla_sample_node:/home/bolty/ament_ws/src/carla_sample_node
+
+  carla_notebooks:
```
[the remark this answers is not in the document]
Comment (Edwardius) [reply modules/dev_overrides/docker-compose.simulation.yaml:35]: you can mount the notebook in the non-override file and move this code there instead.

File: src/simulation/carla_config/src/carla_mpc_bridge.cpp
[the remark this answers is not in the document]
Comment (Edwardius) [reply src/simulation/carla_config/src/carla_mpc_bridge.cpp:1]: ... I'll let this slide .........................

File: src/simulation/carla_config/src/carla_mpc_bridge.cpp [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    } while(role_name == "");
+
+    rclcpp::Subscription<path_planning_msgs::msg::CarlaEgoVehicleStatus>::SharedPtr steeringAngleSub;
+    steeringAngleSub = node->create_subscription<path_planning_msgs::msg::CarlaEgoVehicleStatus>(
+        "/carla/" + role_name + "/vehicle_status", 2, 
+        publish_steering);
+
+    rclcpp::Subscription<path_planning_msgs::msg::MPCOutput>::SharedPtr mpcSub;
+    mpcSub = node->create_subscription<path_planning_msgs::msg::MPCOutput>(
+        "mpc_output", 2,
```
[the remark this answers is not in the document]
Comment (Edwardius) [reply src/simulation/carla_config/src/carla_mpc_bridge.cpp:73]: actually, moreso for me. The selling point is, mpc might create two different models, publishing to two diff topics, carla can switch to either

File: src/simulation/carla_config/src/carla_mpc_bridge.cpp [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    std::string role_name;
+    std::string param_name = "role_name";
+    do{
+        role_name = parameters_client->get_parameters({"role_name"})[0].get_value<std::string>();
+        std::this_thread::sleep_for(std::chrono::seconds(1));
+    } while(role_name == "");
+
+    rclcpp::Subscription<path_planning_msgs::msg::CarlaEgoVehicleStatus>::SharedPtr steeringAngleSub;
+    steeringAngleSub = node->create_subscription<path_planning_msgs::msg::CarlaEgoVehicleStatus>(
+        "/carla/" + role_name + "/vehicle_status", 2,
```
Comment (Edwardius) [inline src/simulation/carla_config/src/carla_mpc_bridge.cpp:68]: generally, please expose out all (as much of) ros topics into a yaml config as possible

File: src/wato_msgs/simulation/common_msgs/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+find_package(geometry_msgs REQUIRED)
+
+set(msg_files
+    msg/Obstacle.msg
+    msg/ParkingSpot.msg
+    msg/LaneLine.msg
+    msg/RoadLinesList.msg
+    msg/StopLine.msg
+    msg/TrafficLight.msg
+    msg/TrafficSign.msg)
```
Comment (Edwardius) [inline src/wato_msgs/simulation/common_msgs/CMakeLists.txt:25]: we're probably deprecating alot of these messages, does CARLA rosbridge convert to these??

**Review by Edwardius — CHANGES_REQUESTED** [review]
Looks awesome! Couple of comments here and there, but it should be ready to merge in no time

File: docker/simulation/carla_sample_node/carla_sample_node.Dockerfile (hunk 3 above)
Comment (VishGit1234) [reply docker/simulation/carla_sample_node/carla_sample_node.Dockerfile:18]: Yeah seems git is already installed on the base image

File: watod_scripts/watod-setup-env.sh [hunk 11]
```diff
@@ -123,6 +127,9 @@ SETGID=$(id -g)
 BASE_PORT=${BASE_PORT:-$(($(id -u)*20))}
 GUI_TOOLS_VNC_PORT=${GUI_TOOLS_VNC_PORT:-$((BASE_PORT++))}
 FOXGLOVE_BRIDGE_PORT=${FOXGLOVE_BRIDGE_PORT:-$((GUI_TOOLS_VNC_PORT++))}
+CARLAVIZ_PORT=${CARLAVIZ_PORT:-$((GUI_TOOLS_VNC_PORT++))}
+CARLAVIZ_PORT_2=${CARLAVIZ_PORT_2:-$((GUI_TOOLS_VNC_PORT++))}
+CARLA_NOTEBOOKS_PORT=${CARLA_NOTEBOOKS_PORT:-$((GUI_TOOLS_VNC_PORT++))}
```
[the remark this answers is not in the document]
Comment (VishGit1234) [reply watod_scripts/watod-setup-env.sh:132]: The setup is specifically for 0.9.13, the ROS bridge only supports that version of CARLA

File: docs/carla.md [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+## Using a ROS Node to interact with CARLA
+
+**Ensure that you have a good understanding of ROS and writing ROS nodes before proceeding**
+
+The currently recommended (and the simplest) way of using a ROS Node to interact with CARLA is to simply edit the already setup CARLA sample node (currently only in Python unfortunately). Since the package (and all required CARLA and ROS dependencies) is already fully setup the only file you need to edit is the `carla_sample_node.py` found under `wato_monorepo_v2/src/simulation/carla_sample_node/carla_sample_node/`. Since the container is setup with a volume, simply restart the containers with `watod down` followed by `watod up` (or any other method to restart the containers) and the node should be rebuilt and launched with the changes you made to `carla_sample_node.py`. The sample node currently shows how to publish a message to enable (and keep enabled) autopilot, subsribe to the GNSS sensor topic and output the data to the console.
```
[the remark this answers is not in the document]
Comment (VishGit1234) [reply docs/carla.md:44]: I should re-word this, the sample node just has all the carla related messages installed

**Comment by Edwardius** [comment]
Is this ready for rereview?

**Comment by VishGit1234** [comment]
> Is this ready for rereview?

Not yet, been a bit busy I'll try and get it done this weekend.

File: docs/carla.md [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+**On your local machine (your personal laptop/pc)**
+
+In any browser go to `localhost:8081` (replace 8081 with the port specified in the `CARLAVIZ_PORT` variable) and after waiting 10 seconds or so CarlaViz should appear.  
+
+## FAQ
+
+### CARLA is running very slow (approx. 3 fps)
+
+This is expected. The ROS bridge causes CARLA to render all the sensor data which slows down the simulation considerably. While this may be annoying when viewing real-time camera output or trying to control the car manual, the simulation is still running accurately.
```
[the remark this answers is not in the document]
Comment (VishGit1234) [reply docs/carla.md:74]: Yep, recording and playback can be easily done using CARLA python API (so via the jupyter notebook container)

File: src/wato_msgs/simulation/common_msgs/CMakeLists.txt (hunk 10 above)
Comment (VishGit1234) [reply src/wato_msgs/simulation/common_msgs/CMakeLists.txt:25]: Nope, was just being used in one of the path planning messages, will get rid of them

File: src/simulation/carla_config/src/carla_mpc_bridge.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (VishGit1234) [reply src/simulation/carla_config/src/carla_mpc_bridge.cpp:73]: Implemented using template function for configurability

File: src/simulation/carla_config/src/carla_mpc_bridge.cpp (hunk 9 above)
Comment (VishGit1234) [reply src/simulation/carla_config/src/carla_mpc_bridge.cpp:68]: I'm not sure what you mean by this, could you go into a little more depth about what you're looking for here?

File: .github/include/base_image_config.json [hunk 14]
```diff
@@ -7,10 +7,13 @@
...
-
-
-
+    {
+        "tag": "foxy-ubuntu20.04",
+        "external_image": "ubuntu:20.04",
+        "injections": {
+            "ros2": "foxy"
+        }
+    },
```
[the remark this answers is not in the document]
Comment (VishGit1234) [reply .github/include/base_image_config.json]: Done

File: .github/workflows/build_base_images.yml [hunk 15]
```diff
@@ -113,4 +113,4 @@ jobs:
           target: wato_base
           build-args: | 
             GENERIC_IMAGE=${{ steps.construct-registry-url.outputs.url }}_generic
-            USER_PASSWD=${{ secrets.MONOREPO_PASSWD }}
+            USER_PASSWD=${{ secrets.MONOREPO_PASSWD }}
```
[the remark this answers is not in the document]
Comment (VishGit1234) [reply .github/workflows/build_base_images.yml]: Sorry that was by accident, there should be no change. Should be fixed now

File: docker/simulation/carla_ros_bridge/carla_ros_bridge.Dockerfile [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+#Install Python Carla API
+COPY --from=wato_carla_api --chown=root /home/carla/PythonAPI/carla /opt/carla/PythonAPI
+WORKDIR /opt/carla/PythonAPI
+RUN python3.8 -m easy_install pip && \
+      pip3 install carla==${CARLA_VERSION} && \
+      pip install simple-pid==2.0.0 && \
+      pip install transforms3d==0.4.1 && \
+      pip install pexpect==4.9.0 && \
+      pip install networkx==3.1
+
```
[the remark this answers is not in the document]
Comment (VishGit1234) [reply docker/simulation/carla_ros_bridge/carla_ros_bridge.Dockerfile]: Done.

File: docker/simulation/carla_ros_bridge/carla_ros_bridge.Dockerfile [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+WORKDIR ${AMENT_WS}/src
+
+# Download ROS Bridge
+RUN git clone --depth 1 --branch master --recurse-submodules https://github.com/carla-simulator/ros-bridge.git
+
+# Download rviz 
+RUN git clone --branch foxy --recurse-submodules https://github.com/ros2/rviz.git
+
+# Fix an error in the ackermann_control node 
+RUN sed -i s/simple_pid.PID/simple_pid.pid/g ./ros-bridge/carla_ackermann_control/src/carla_ackermann_control/carla_ackermann_control_node.py
```
[the remark this answers is not in the document]
Comment (VishGit1234) [reply docker/simulation/carla_ros_bridge/carla_ros_bridge.Dockerfile]: Done.

File: docker/simulation/carla_ros_bridge/carla_ros_bridge.Dockerfile (hunk 1 above)
Comment (VishGit1234) [reply docker/simulation/carla_ros_bridge/carla_ros_bridge.Dockerfile]: Removed curl and git from installation

File: docker/simulation/carla_sample_node/carla_sample_node.Dockerfile (hunk 2 above)
Comment (VishGit1234) [reply docker/simulation/carla_sample_node/carla_sample_node.Dockerfile]: Removed COPY sample_msgs

File: docs/carla.md [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+    - [Interacting with CARLA using the Python API](#interacting-with-carla-using-the-python-api)
+    - [Using a ROS Node to interact with CARLA](#using-a-ros-node-to-interact-with-carla)
+    - [CARLA Visualization using Foxglove Studio](#carla-visualization-using-foxglove-studio)
+    - [CARLA Visualization using CarlaViz](#carla-visualization-using-carlaviz)
+    - [FAQ](#faq)
+        - [CARLA is running very slow (approx. 3 fps)](#carla-is-running-very-slow-approx-3-fps)
+
+
+## Getting Started
+**READ THIS**: [docs/setup.md](docs/setup.md)
```
[the remark this answers is not in the document]
Comment (VishGit1234) [reply docs/carla.md]: Removed.

File: docs/carla.md [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+
+**READ THIS**: [docs/readme.md](docs/readme.md)
```
[the remark this answers is not in the document]
Comment (VishGit1234) [reply docs/carla.md]: Removed.

File: modules/dev_overrides/docker-compose.simulation.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (VishGit1234) [reply modules/dev_overrides/docker-compose.simulation.yaml:35]: Done.

File: src/simulation/carla_config/src/carla_mpc_bridge.cpp [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+    mpcSub = node->create_subscription<path_planning_msgs::msg::MPCOutput>(
+        "mpc_output", 2, 
+        mpc_to_carla);
+    
+    rclcpp::Subscription<path_planning_msgs::msg::Environment>::SharedPtr envSub;
+    envSub = node->create_subscription<path_planning_msgs::msg::Environment>(
+        "/path_planning/environment", 2, 
+        env_callback);
+
+    steeringPub = node->create_publisher<embedded_msgs::msg::SteeringAngleCAN>("steering_data", 1);
```
[the remark this answers is not in the document]
Comment (VishGit1234) [reply src/simulation/carla_config/src/carla_mpc_bridge.cpp]: Done

File: src/wato_msgs/simulation/path_planning_msgs/CMakeLists.txt
[the remark this answers is not in the document]
Comment (VishGit1234) [reply src/wato_msgs/simulation/path_planning_msgs/CMakeLists.txt:1]: I think the naming scheme is a bit confusing, but I think these messages still belong under simulation

**Comment by VishGit1234** [comment]
Done addressing the PR comments @Edwardius . Please take a look when you have a chance

**Review by Edwardius — APPROVED** [review]
LGTM, make the small changes, resolve conflict, and ur good to go!

File: docker/simulation/carla_ros_bridge/carla_ros_bridge.Dockerfile [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+RUN git clone --depth 1 --branch master --recurse-submodules https://github.com/carla-simulator/ros-bridge.git && \
+    cd ros-bridge && \
+    git checkout e9063d97ff5a724f76adbb1b852dc71da1dcfeec && \
+    cd ..
+
+# Download rviz 
+RUN git clone --depth 1 --branch foxy --recurse-submodules https://github.com/ros2/rviz.git && \
+    cd rviz && \
+    git checkout 22179fc7a2fed72cd3f38e6712c62e2228a36001 && \
+    cd ..
```
[the remark this answers is not in the document]
Comment (VishGit1234) [reply docker/simulation/carla_ros_bridge/carla_ros_bridge.Dockerfile:21]: Unfortunately don't think I can do that for the ROS bridge, they haven't created a release for 0.9.13.

File: docker/simulation/carla_sample_node/carla_sample_node.Dockerfile [hunk 22]
```diff
@@ -6,8 +6,26 @@ FROM ${BASE_IMAGE} as source
...
+
+RUN git clone --depth 1 https://github.com/ros-perception/image_common.git --branch $ROS_DISTRO && \
+    cd image_common && \
+    git checkout e947b47a45971e3edb59d8e34bc8e7cd2a41f2e6 && \
+    cd ..
+
+RUN git clone --depth 1 https://github.com/carla-simulator/ros-carla-msgs.git --branch master && \
+    cd ros-carla-msgs && \
+    git checkout 081fdcdc8a4f8d4ce936b792c4ef13aba3fd7478 && \
+    cd ..
```
[the remark this answers is not in the document]
Comment (VishGit1234) [reply docker/simulation/carla_sample_node/carla_sample_node.Dockerfile:28]: Some of these are specific branches and don't have corresponding releases, so I'll keep those ones specific commit hashes

**Pull request merged**